dxp/react_rendering.md

```diff
@@ -0,0 +1,148 @@
+# React Rendering in DXP
+
+This article describes DXP infrastructure to render React components from the server side perspective. This includes taglibs as well as low-level services.
+
+## React Renderer Service
+
+-   OSGi bundle: `com.liferay.portal.template.react.renderer.api`
+-   Interface name: `com.liferay.portal.template.react.renderer.ReactRenderer`
+-   Service type: `singleton`
+
+This is the low-level service to render React components.
+
+> Note that it renders React components, not any React code, so you must provide a source which exports a React component to render.
+
+It has one method:
+
+```java
+public void renderReact(
+		ComponentDescriptor componentDescriptor, Map<String, Object> data,
+		HttpServletRequest httpServletRequest, Writer writer)
+	throws IOException;
+```
+
+which receives:
+
+1. A `ComponentDescriptor` describing the source of the React component to render. More on this in the next section.
+2. A `Map` containing arbitrary data that will be injected to the React component as initial properties. Note that the data inside the map must be compatible with JavaScript, so you cannot inject any Java object you like, only primitive types, `Map`s, `Array`s, ... that can be serialized to JSON.
+3. The current `HttpServletRequest` for which the component is being rendered.
+4. A `Writer` where the rendered HTML (and optionally the JavaScript) is written to.
+
+Note that parameter 4 is not intended to be used to write the rendered HTML and JavaScript to any place you like but to pass the `PrintWriter` from the `HttpServletResponse` or the `JspWriter` from the `PageContext` you have available when invoking the method.
+
+Also, take into account that, if you don't render the JavaScript inline (see option `positionInline` of `ComponentDescriptor`), the JavaScript will be written to the bottom of the current `HttpServletResponse`, while the HTML will be inlined into the passed `Writer`.
```

Comment:
It would be interesting to link to what `HTML` is actually being rendered and why. That is, the `div` placeholder that we use to mount the component.

I get the impression from reading here that this can render arbitrary HTML (which might in the future if we server-render the components, but it's not the case now...)

Reply:
It's described some lines below.